### PR TITLE
De-duplicate some code on control plane bootstrap and join

### PR DESCRIPTION
**Pull request by louiseschmidtgen**
## Description
This PR reduces the code duplication in `hooks_bootstrap.go` and `hooks_join.go` and puts common functionality in `cluster_util.go`.

As a bonus this PR also improves some naming and error messages.

Note: Combining Certificates code is not possible as it is two seperate code paths.

### CODE REVIEW

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		}
+	default:
+		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores)
+	}
+
+	return nil
+
+}
+
+func generateKubeconfigs(snap snap.Snap, s *state.State, cfg types.ClusterConfig) error {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:56]: Reduce what is passed here, and adjust name to better reflect what the function does without having to lookup the implementation
```suggestion
func setupKubeconfigs(ctx context.Context, snap snap.Snap, securePort int, caCert string) error {
```

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	"github.com/canonical/k8s/pkg/k8sd/pki"
+	"github.com/canonical/k8s/pkg/k8sd/setup"
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/canonical/microcluster/state"
+)
+
+func generateAndEnsureCertificates(snap snap.Snap, cfg types.ClusterConfig, nodeName string, allowSelfSignedCA bool) error {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:18]: also update caller sites

```suggestion
func setupDatastoreCertificates(snap snap.Snap, cfg types.ClusterConfig, nodeName string, allowSelfSignedCA bool) error {
```

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 3]
```diff
@@ -178,37 +176,9 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
-			return fmt.Errorf("failed to initialize cluster certificates: %w", err)
-		}
-		if err := setup.EnsureExtDatastorePKI(snap, certificates); err != nil {
-			return fmt.Errorf("failed to write cluster certificates: %w", err)
-		}
-	default:
-		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores)
+	err = generateAndEnsureCertificates(snap, cfg, s.Name(), true)
+	if err != nil {
+		return fmt.Errorf("failed to generate and ensure certificates: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:181]: inline check err 
```suggestion
		if err := generateAndEnsureCertificates(snap, cfg, s.Name(), true)
; err != nil {
		return fmt.Errorf("failed to generate and ensure certificates: %w", err)
```

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 4]
```diff
@@ -236,24 +206,9 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
-		token, err := impl.GetOrCreateAuthToken(s.Context, s, kubeconfig.username, kubeconfig.groups)
-		if err != nil {
-			return fmt.Errorf("failed to generate token for username=%s groups=%v: %w", kubeconfig.username, kubeconfig.groups, err)
-		}
-		if err := setup.Kubeconfig(path.Join(snap.KubernetesConfigDir(), kubeconfig.file), token, fmt.Sprintf("127.0.0.1:%d", cfg.APIServer.SecurePort), cfg.Certificates.CACert); err != nil {
-			return fmt.Errorf("failed to write kubeconfig %s: %w", kubeconfig.file, err)
-		}
+	err = generateKubeconfigs(snap, s, cfg)
+	if err != nil {
+		return err
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:211]: inline error checks 
```suggestion
		if err = generateKubeconfigs(snap, s, cfg)
; err != nil {
		return err
```

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 5]
```diff
@@ -267,25 +222,9 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
-	}
-	if err := setup.KubeScheduler(snap); err != nil {
-		return fmt.Errorf("failed to configure kube-scheduler: %w", err)
-	}
-
-	if err := setup.KubeAPIServer(snap, cfg.Network.ServiceCIDR, s.Address().Path("1.0", "kubernetes", "auth", "webhook").String(), true, cfg.APIServer.Datastore, cfg.APIServer.DatastoreURL, cfg.APIServer.AuthorizationMode); err != nil {
-		return fmt.Errorf("failed to configure kube-apiserver: %w", err)
+	err = configureServicesControlPlane(snap, s, cfg, nodeIP)
+	if err != nil {
+		return fmt.Errorf("failed to configure services: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:227]: same here

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 6]
```diff
@@ -299,28 +238,9 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
-	// Wait for API server to come up
-	client, err := k8s.NewClient(snap)
+	err = startServicesControlPlane(snap, s, cfg)
 	if err != nil {
-		return fmt.Errorf("failed to create k8s client: %w", err)
-	}
-
-	if err := client.WaitApiServerReady(s.Context); err != nil {
-		return fmt.Errorf("k8s api server did not become ready in time: %w", err)
+		return fmt.Errorf("failed to start services: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:243]: same here

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 7]
```diff
@@ -40,35 +37,9 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
-			return fmt.Errorf("failed to initialize cluster certificates: %w", err)
-		}
-		if err := setup.EnsureExtDatastorePKI(snap, certificates); err != nil {
-			return fmt.Errorf("failed to write cluster certificates: %w", err)
-		}
-	default:
-		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores)
+	err = generateAndEnsureCertificates(snap, cfg, s.Name(), false)
+	if err != nil {
+		return fmt.Errorf("failed to generate and ensure certificates: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:42]: same here

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 8]
```diff
@@ -95,25 +66,9 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
-		token, err := impl.GetOrCreateAuthToken(s.Context, s, kubeconfig.username, kubeconfig.groups)
-		if err != nil {
-			return fmt.Errorf("failed to generate token for username=%s groups=%v: %w", kubeconfig.username, kubeconfig.groups, err)
-		}
-		if err := setup.Kubeconfig(path.Join(snap.KubernetesConfigDir(), kubeconfig.file), token, fmt.Sprintf("127.0.0.1:%d", cfg.APIServer.SecurePort), cfg.Certificates.CACert); err != nil {
-			return fmt.Errorf("failed to write kubeconfig %s: %w", kubeconfig.file, err)
-		}
+	err = generateKubeconfigs(snap, s, cfg)
+	if err != nil {
+		return fmt.Errorf("failed to generate kubeconfigs: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:71]: same here

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 9]
```diff
@@ -140,50 +95,15 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
-	}
-	if err := setup.KubeScheduler(snap); err != nil {
-		return fmt.Errorf("failed to configure kube-scheduler: %w", err)
-	}
-
-	if err := setup.KubeAPIServer(snap, cfg.Network.ServiceCIDR, s.Address().Path("1.0", "kubernetes", "auth", "webhook").String(), true, cfg.APIServer.Datastore, cfg.APIServer.DatastoreURL, cfg.APIServer.AuthorizationMode); err != nil {
-		return fmt.Errorf("failed to configure kube-apiserver: %w", err)
+	err = configureServicesControlPlane(snap, s, cfg, nodeIP)
+	if err != nil {
+		return fmt.Errorf("failed to configure services: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:100]: same here

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 10]
```diff
@@ -140,50 +95,15 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
-	client, err := k8s.NewClient(snap)
+	err = startServicesControlPlane(snap, s, cfg)
 	if err != nil {
-		return fmt.Errorf("failed to create kubernetes client: %w", err)
-	}
-
-	if err := client.WaitApiServerReady(s.Context); err != nil {
-		return fmt.Errorf("kube-apiserver did not become ready: %w", err)
+		return fmt.Errorf("failed to start services: %w", err)
 	}
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:107]: same here

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+		}
+		if err := setup.Kubeconfig(path.Join(snap.KubernetesConfigDir(), kubeconfig.file), token, fmt.Sprintf("127.0.0.1:%d", cfg.APIServer.SecurePort), cfg.Certificates.CACert); err != nil {
+			return fmt.Errorf("failed to write kubeconfig %s: %w", kubeconfig.file, err)
+		}
+	}
+	return nil
+
+}
+
+func configureServicesControlPlane(snap snap.Snap, s *state.State, cfg types.ClusterConfig, nodeIP net.IP) error {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:81]:
```suggestion
func setupControlPlaneServices(snap snap.Snap, s *state.State, cfg types.ClusterConfig, nodeIP net.IP) error {
```

**Review by neoaggelos — COMMENTED** [review]
Very nice work @louiseschmidtgen! Leaving some comments regarding structure.

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 12]
```diff
@@ -178,37 +176,9 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
-		})
-		if err := certificates.CompleteCertificates(); err != nil {
-			return fmt.Errorf("failed to initialize cluster certificates: %w", err)
-		}
-		if err := setup.EnsureK8sDqlitePKI(snap, certificates); err != nil {
-			return fmt.Errorf("failed to write cluster certificates: %w", err)
-		}
-
-		cfg.Certificates.K8sDqliteCert = certificates.K8sDqliteCert
-		cfg.Certificates.K8sDqliteKey = certificates.K8sDqliteKey
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:197]: NOTE: this is now missing in the bootstrap code, which is also why the e2e tests are failing. perhaps you can add it after the `setupDatastoreCertificates`?

File: src/k8s/pkg/k8sd/app/cluster_util.go (hunk 1 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/k8sd/app/cluster_util.go]: Couldn't get rid of passing state here but I also reduced what gets passed for the other functions.

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go (hunk 12 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/k8sd/app/hooks_bootstrap.go:197]: Ah thanks for catching that for me!

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	if err := setup.KubeScheduler(snap); err != nil {
+		return fmt.Errorf("failed to configure kube-scheduler: %w", err)
+	}
+	if err := setup.KubeAPIServer(snap, cfg.Network.ServiceCIDR, s.Address().Path("1.0", "kubernetes", "auth", "webhook").String(), true, cfg.APIServer.Datastore, cfg.APIServer.DatastoreURL, cfg.APIServer.AuthorizationMode); err != nil {
+		return fmt.Errorf("failed to configure kube-apiserver: %w", err)
+	}
+	return nil
+}
+
+func startServicesControlPlane(ctx context.Context, snap snap.Snap, datastore string) error {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:103]:
```suggestion
func startControlPlaneServices(ctx context.Context, snap snap.Snap, datastore string) error {
```

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+	case "k8s-dqlite":
+		dqliteCert := pki.NewK8sDqlitePKI(pki.K8sDqlitePKIOpts{
+			Hostname:          nodeName,
+			IPSANs:            []net.IP{{127, 0, 0, 1}},
+			Years:             20,
+			AllowSelfSignedCA: allowSelfSignedCA,
+		})
+
+		if err := dqliteCert.CompleteCertificates(); err != nil {
+			return fmt.Errorf("failed to initialize cluster certificates: %w", err), nil, nil
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:31]: let's update the error messages here (and in other places) to reflect what the code is trying to do

```suggestion
			return fmt.Errorf("failed to initialize k8s-dqlite certificates: %w", err), nil, nil
```

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		if err := setup.EnsureK8sDqlitePKI(snap, dqliteCert); err != nil {
+			return fmt.Errorf("failed to write cluster certificates: %w", err), nil, nil
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:34]:
```suggestion
			return fmt.Errorf("failed to write k8s-dqlite certificates: %w", err), nil, nil
```

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		return nil, dqliteCert, nil
+	case "external":
+		externalDatastoreCert := &pki.ExternalDatastorePKI{
+			DatastoreCACert:     cfg.Certificates.DatastoreCACert,
+			DatastoreClientCert: cfg.Certificates.DatastoreClientCert,
+			DatastoreClientKey:  cfg.Certificates.DatastoreClientKey,
+		}
+		if err := externalDatastoreCert.CheckCertificates(); err != nil {
+			return fmt.Errorf("failed to initialize cluster certificates: %w", err), nil, nil
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:44]: same here

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/canonical/k8s/pkg/k8sd/pki"
+	"github.com/canonical/k8s/pkg/k8sd/setup"
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/canonical/microcluster/state"
+)
+
+func setupDatastoreCertificates(snap snap.Snap, cfg types.ClusterConfig, nodeName string, allowSelfSignedCA bool) (error, *pki.K8sDqlitePKI, *pki.ExternalDatastorePKI) {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:19]: this is not a useful function, as the semantics are blurred. first of all, the error ought to be the last return value. then, we return two distinct pkis, and we don't define when either will be nil.

this leads me to think that perhaps we should avoid this completely. if these are two separate code-paths in bootstrap and join, then let's treat them as such and not group them like this.

File: src/k8s/pkg/k8sd/app/cluster_util.go (hunk 13 above)
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:103]: This function does not only start the control plane services, it also waits for kube-apiserver to become ready.

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+	// Wait for API server to come up
+	client, err := k8s.NewClient(snap)
+	if err != nil {
+		return fmt.Errorf("failed to create k8s client: %w", err)
+	}
+
+	if err := client.WaitApiServerReady(ctx); err != nil {
+		return fmt.Errorf("k8s api server did not become ready in time: %w", err)
+	}
+
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:128]: better split that into a separate function? or rename to `startControlPlaneServicesAndWaitReady(....)`

In general, all function names should declare exactly what they do, without callers having to look up the implementation for details.

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 19]
```diff
@@ -40,80 +37,40 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
-		}
-		if err := certificates.CheckCertificates(); err != nil {
-			return fmt.Errorf("failed to initialize cluster certificates: %w", err)
-		}
-		if err := setup.EnsureExtDatastorePKI(snap, certificates); err != nil {
-			return fmt.Errorf("failed to write cluster certificates: %w", err)
-		}
-	default:
-		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores)
+	// TODO check if join node gets selfsigned cert allowed
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:40]: no, all k8s-dqlite nodes must use the same certificates. see my other comment, perhaps it's best to keep the certificate generation as is for now.

**Review by neoaggelos — CHANGES_REQUESTED** [review]
Looks nicer, did another round of comments

File: src/k8s/pkg/k8sd/app/cluster_util.go (hunk 14 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/k8sd/app/cluster_util.go:34]: I was beginning to think the same, this step looks the same on the first glance but actually isn't.

File: src/k8s/pkg/k8sd/app/cluster_util.go (hunk 17 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/k8sd/app/cluster_util.go]: I was beginning to think the same, this section looks the same on the first glance but actually isn't.

File: src/k8s/pkg/k8sd/app/hooks_join.go (hunk 19 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/k8sd/app/hooks_join.go]: Agreed it was getting way to messy, I left them as is now.

File: src/k8s/pkg/k8sd/app/cluster_util.go (hunk 18 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/k8sd/app/cluster_util.go:95]: Yes, I think two functions is better indeed.

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+		return fmt.Errorf("unsupported datastore %s, must be one of %v", datastore, setup.SupportedDatastores)
+	}
+
+	if err := snaputil.StartControlPlaneServices(ctx, snap); err != nil {
+		return fmt.Errorf("failed to start control plane services: %w", err)
+	}
+	return nil
+}
+
+func waitReadyApiServer(ctx context.Context, snap snap.Snap) error {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:84]: this is `waitReadyApiServer` and it calls `WaitApiServerReady`. Using different names for similar things quickly messes things up in a codebase, let's call this `waitApiServerReady`

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+
+	// Wait for API server to come up
+	client, err := k8s.NewClient(snap)
+	if err != nil {
+		return fmt.Errorf("failed to create k8s client: %w", err)
+	}
+
+	if err := client.WaitApiServerReady(ctx); err != nil {
+		return fmt.Errorf("k8s api server did not become ready in time: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:93]: The service name is kube-apiserver
```suggestion
		return fmt.Errorf("kube-apiserver did not become ready in time: %w", err)
```

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	return nil
+}
+
+func waitReadyApiServer(ctx context.Context, snap snap.Snap) error {
+
+	// Wait for API server to come up
+	client, err := k8s.NewClient(snap)
+	if err != nil {
+		return fmt.Errorf("failed to create k8s client: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/cluster_util.go:89]:
```suggestion
		return fmt.Errorf("failed to create Kubernetes client: %w", err)
```

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 23]
```diff
@@ -101,23 +99,23 @@ func onBootstrapWorkerNode(s *state.State, encodedToken string) error {
 	}
 
 	// Certificates
-	certificates := &pki.WorkerNodePKI{
+	workerNodeCert := &pki.WorkerNodePKI{
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:102]: in this context, `workerNodeCert` does not help more than `certificates`, let's keep the other name.

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 24]
```diff
@@ -101,23 +99,23 @@ func onBootstrapWorkerNode(s *state.State, encodedToken string) error {
...
 		KubeletKey:  response.KubeletKey,
 	}
-	if err := certificates.CompleteCertificates(); err != nil {
-		return fmt.Errorf("failed to initialize cluster certificates: %w", err)
+	if err := workerNodeCert.CompleteCertificates(); err != nil {
+		return fmt.Errorf("failed to initialize worker node certificates: %w", err)
 	}
-	if err := setup.EnsureWorkerPKI(snap, certificates); err != nil {
-		return fmt.Errorf("failed to write cluster certificates: %w", err)
+	if err := setup.EnsureWorkerPKI(snap, workerNodeCert); err != nil {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:110]: See here how this `setup.EnsureWorkerPKI(snap, workerNodeCert)` "stutters"

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 25]
```diff
@@ -180,80 +178,65 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
 
 	switch cfg.APIServer.Datastore {
 	case "k8s-dqlite":
-		certificates := pki.NewK8sDqlitePKI(pki.K8sDqlitePKIOpts{
+		dqliteCert := pki.NewK8sDqlitePKI(pki.K8sDqlitePKIOpts{
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:181]: same here, this is not needed

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 26]
```diff
@@ -180,80 +178,65 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
 		}
 
-		cfg.Certificates.K8sDqliteCert = certificates.K8sDqliteCert
-		cfg.Certificates.K8sDqliteKey = certificates.K8sDqliteKey
+		cfg.Certificates.K8sDqliteCert = dqliteCert.K8sDqliteCert
+		cfg.Certificates.K8sDqliteKey = dqliteCert.K8sDqliteKey
+
 	case "external":
-		certificates := &pki.ExternalDatastorePKI{
+		externalDatastoreCert := &pki.ExternalDatastorePKI{
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:198]: same here

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 27]
```diff
@@ -180,80 +178,65 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
+		if err := setup.EnsureExtDatastorePKI(snap, externalDatastoreCert); err != nil {
+			return fmt.Errorf("failed to write external datastore certificates: %w", err)
 		}
 	default:
 		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores)
 	}
 
 	// Certificates
-	certificates := pki.NewControlPlanePKI(pki.ControlPlanePKIOpts{
+	controlPlaneCert := pki.NewControlPlanePKI(pki.ControlPlanePKIOpts{
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:214]: same here

File: src/k8s/pkg/k8sd/app/hooks_bootstrap.go [hunk 28]
```diff
@@ -299,28 +266,13 @@ func onBootstrapControlPlane(s *state.State, initConfig map[string]string) error
...
-		return fmt.Errorf("failed to create k8s client: %w", err)
+	if err := startControlPlaneServices(s.Context, snap, cfg.APIServer.Datastore); err != nil {
+		return fmt.Errorf("failed to start services: %w", err)
 	}
 
-	if err := client.WaitApiServerReady(s.Context); err != nil {
-		return fmt.Errorf("k8s api server did not become ready in time: %w", err)
+	// Wait until Kube-API server is ready
+	if err := waitReadyApiServer(s.Context, snap); err != nil {
+		return fmt.Errorf("failed to wait for Kube-API server: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_bootstrap.go:275]: Let's be consistent with `kube-apiserver` in the error messages, using different similar names in different places causes confusion.

```suggestion
	if err := waitReadyApiServer(s.Context, snap); err != nil {
		return fmt.Errorf("failed to wait for kube-apiserver: %w", err)
```

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 29]
```diff
@@ -40,80 +37,64 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
 		return fmt.Errorf("failed to get IP address(es) from ServiceCIDR %q: %w", cfg.Network.ServiceCIDR, err)
 	}
 
+	// Certificates
 	switch cfg.APIServer.Datastore {
 	case "k8s-dqlite":
-		certificates := pki.NewK8sDqlitePKI(pki.K8sDqlitePKIOpts{
+		dqliteCert := pki.NewK8sDqlitePKI(pki.K8sDqlitePKIOpts{
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:43]: same here

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 30]
```diff
@@ -40,80 +37,64 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
+			return fmt.Errorf("failed to initialize k8s-dqlite certificates: %w", err)
 		}
-		if err := setup.EnsureK8sDqlitePKI(snap, certificates); err != nil {
-			return fmt.Errorf("failed to write cluster certificates: %w", err)
+		if err := setup.EnsureK8sDqlitePKI(snap, dqliteCert); err != nil {
+			return fmt.Errorf("failed to write k8s-dqlite certificates: %w", err)
 		}
 	case "external":
-		certificates := &pki.ExternalDatastorePKI{
+		externalDatastoreCert := &pki.ExternalDatastorePKI{
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:57]: same here

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 31]
```diff
@@ -40,80 +37,64 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
+		if err := setup.EnsureExtDatastorePKI(snap, externalDatastoreCert); err != nil {
+			return fmt.Errorf("failed to write external datastore certificates: %w", err)
 		}
 	default:
 		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores)
 	}
 
 	// Certificates
-	certificates := pki.NewControlPlanePKI(pki.ControlPlanePKIOpts{
+	controlPlaneCert := pki.NewControlPlanePKI(pki.ControlPlanePKIOpts{
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:73]: same here

File: src/k8s/pkg/k8sd/app/hooks_join.go [hunk 32]
```diff
@@ -141,49 +122,18 @@ func onPostJoin(s *state.State, initConfig map[string]string) error {
...
-		return fmt.Errorf("failed to create kubernetes client: %w", err)
+	if err := startControlPlaneServices(s.Context, snap, cfg.APIServer.Datastore); err != nil {
+		return fmt.Errorf("failed to start services: %w", err)
 	}
 
-	if err := client.WaitApiServerReady(s.Context); err != nil {
-		return fmt.Errorf("kube-apiserver did not become ready: %w", err)
+	// Wait until Kube-API server is ready
+	if err := waitReadyApiServer(s.Context, snap); err != nil {
+		return fmt.Errorf("failed to wait for Kube-API server: %w", err)
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/app/hooks_join.go:136]: same here

**Review by neoaggelos — APPROVED** [review]

**Pull request merged**